core/nightshadev2/src/nightshade.rs

```diff
@@ -0,0 +1,244 @@
+use std::collections::HashSet;
+use std::cmp::Ordering;
+
+pub type AuthorityId = usize;
+pub type BLSSignature = u64;
+
+pub enum NSResult{
+    Updated(Option<State>),
+    Error(String),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+struct BareState {
```

Comment:
We can derive `PartialOrd` and `Ord` then they will produce lexicographic ordering which is what we want, see https://doc.rust-lang.org/std/cmp/trait.Ord.html#derivable

Reply:
Actually, the way fields are declared (relative order and signs), lexicographic order doesn't work. I think is better if we keep this order (in order to have more legible code) than hacking this struct so default order works.

Reply:
Can we declare the fields in the correct order? I do not really think the order of declaration of the fields in a struct changes the readability.

Reply:
Correct order should be:

```
struct BareState {
	    confidence0: i64,
            endorses: AuthorityId,
	    confidence1: i64,
}
```
We take greater confidence0 first, lower endorsement for first tie breaker and higher confidence1 for second tie breaker. It is fine using higher endorsement instead of lower, (this is important because otherwise lexocographical order wouldn't fit our case).

Reply:
Fixed [here](https://github.com/nearprotocol/nearcore/pull/549/commits/4e834e54d7f9388b118295caa3bb272f77d8f367)